arch/cortex-m/src/lib.rs

```diff
@@ -15,6 +15,20 @@ pub mod support;
 pub mod syscall;
 pub mod systick;
 
+pub trait CortexMSubarch {
```

Comment:
I think you should leave this commment from the PR description as a doc comment on this trait:

> The fact that some functions are proper trait functions, while others are exposed via associated constants is caused by the associated const functions being #[naked]. To wrap these functions in proper trait functions would require these trait functions to also be #[naked] to avoid generating a function prologue and epilogue, and we'd have to call the respective backing function from within an asm! block. The associated constants allow us to simply reference any function in scope and export it through the provided CortexMSubarch trait infrastructure.

Reply:
Absolutely.

Reply:
I've added some documentation which hopefully explains this trait's purpose and the particular structure chosen here. I've furthermore moved the function documentation comments from the generic ARMv7-M `cortexm` functions into the trait, as this is what should be specifying the function requirements and is rendered as part of the rustdoc.

Reply:
Can we also capture a version of


> This has some significant benefits over passing functions via symbols:

> By using a trait carrying proper first-level Rust functions, the type signatures of the trait and implementing functions are properly validated. Before these changes, some Cortex-M variants already used incorrect type signatures (e.g. *mut u8 instead of *const usize) for the user_stack argument. It also ensures that all functions are provided by a respective sub-architecture at compile time, instead of throwing linker errors.

> Determining the respective functions at compile time, Rust might be able to perform more aggressive inlining, especially if more device-specific proper Rust functions (non hardware-exposed symbols, i.e. not fault or interrupt handlers) were to be added.

> Most importantly, this avoid ambiguity with respect to a compiler fence being inserted by the compiler around calls to switch_to_user. The asm! macro in that function call will cause Rust to emit a compiler fence given the nomem option is not passed, but the opaque extern "C" function call obscured that code path. While this is probably fine and Rust is obliged to generate a compiler fence when switching to C code, having a traceable code path for Rust to the asm! macro will remove any remaining ambiguity and allow us to argue against requiring volatile accesses to userspace memory (during context switches). See https://github.com/tock/tock/pull/2582 for further discussion of this issue.


here, perhaps as a normal `//` comment? This explanation is great to capture independent of github.